feat(sdk): Allow to define and query tables and rows

pascal-botpress opened the pull request:
Tables can be defined this way:

![image](https://github.com/user-attachments/assets/8271866f-bcac-43db-bf67-8a3b23667331)

Rows can be queried this way:

![image](https://github.com/user-attachments/assets/8a1f1cbf-8982-4b49-bc55-0af0bbced022)

pascal-botpress committed:
implement base typings

pascal-botpress committed:
add support for botId in ApiClient

pascal-botpress committed:
add validation schema for tables

pascal-botpress committed:
deploy tables on bp deploy

pascal-botpress committed:
implement response typings for tables

franklevasseur commented on packages/cli/src/command-implementations/deploy-command.ts:
```diff
@@ -348,10 +348,117 @@ export class DeployCommand extends ProjectCommand<DeployCommandDefinition> {
+            'You can delete columns from the studio if you no longer need them.'
+        )
+
+        // TODO: ask the user whether this is a rename. If it is a rename, list
```
**note:** yes, maybe we should think about a common DX for more advanced table schema manipulation

pascal-botpress replied:
Agreed. For now, we might want to merge as-is, but we'll want to invest more time into this in the future.

franklevasseur commented on packages/cli/src/utils/tables/parsing-utils.ts:
**chore:** I'm not sure about where this module should be located, but I don't think `utils` is the correct place since it contains logic that is very botpress specific. I think of the utils folder as a kind of `yadash` (yet another lodash). I think I should be at `packages/cli/src/tables/schemas.ts` or similar

franklevasseur commented on packages/cli/src/utils/tables/parsing-utils.ts:
```diff
@@ -0,0 +1,93 @@
+import * as sdk from '@botpress/sdk'
+
+// FIXME: The table column schemas are defined in @bpinternal/tables-api, but
```
**question:** where is this lib located (`@bpinternal/tables-api`) ?

pascal-botpress replied:
that would be https://github.com/botpress/skynet/blob/1680917/packages/tables-api/src/utils/schema/allowed-schema.ts

franklevasseur commented on packages/cli/src/command-implementations/deploy-command.ts:
```diff
@@ -348,10 +348,117 @@ export class DeployCommand extends ProjectCommand<DeployCommandDefinition> {
+
+    const existingColumns = existingTable.schema.properties
+    const updatedColumns = utils.tables.parsing.columnsSchema.parse(
+      (updatedTableDef.schema.toJsonSchema() as sdk.JSONSchemaOfType<'object'>).properties
```
**nitpick:** maybe dont use this type exported from zui then sdk. I don't know much about it and I will certainly remove it in the next zui major

pascal-botpress replied:
I used `JSONSchemaOfType<'object'>` because I need the underlying `ObjectSchema`, but it's not exported by the module. I guess I could manually redefine the one property I need (`properties`):

```typescript
updatedTableDef.schema.toJsonSchema() as sdk.JSONSchema & {
  properties: {
    [key: string]: sdk.JSONSchema
  }
}
```

franklevasseur commented on packages/cli/src/command-implementations/deploy-command.ts:
```diff
@@ -348,10 +348,117 @@ export class DeployCommand extends ProjectCommand<DeployCommandDefinition> {
+    }
+
+    const existingColumns = existingTable.schema.properties
+    const updatedColumns = utils.tables.parsing.columnsSchema.parse(
```
**question:** if this fails, what error does the CLI user see?

pascal-botpress replied:
Good catch; the user would see a generic zod/zui parsing error. I could improve this by parsing columns one by one in a loop (instead of the whole record in a single pass) and throw an error that says explicitly which column of which table is erroneous.

pascal-botpress committed:
restrict typings

pascal-botpress committed:
make it explicit that rows are read-only

franklevasseur commented on packages/sdk/src/utils/type-utils.ts:
```diff
@@ -5,6 +5,12 @@ export type SafeCast<T, U> = [T] extends [never] ? U : Cast<T, U>
 export type Default<T, U> = undefined extends T ? U : T
 
+export type AtLeastOne<T> = [T, ...T[]]
+export type AtLeastOneProperty<T, U = { [K in keyof T]: Pick<T, K> }> = Partial<T> & U[keyof U]
```
**chore:** these types are quite complex; I think they should be tested in `type-utils.test.ts`. We must ensure that this type is fast to compute

franklevasseur commented on packages/sdk/src/bot/client/types.ts:
**note:** I haven't reviewed this file yet, but I see you are still pushing commits to it so I'll review it later

franklevasseur commented on packages/cli/src/code-generation/bot-implementation/bot-typings/index.ts:
**chore:** we'll probably need the same work done in `code-generation/plugin-implementation/*`

franklevasseur reviewed (commented):
Good job man, this looks so nice. Here's a few problems I encountered when trying it out:

1. filter asks for all columns to be specified:
![Screenshot 2025-01-27 at 10 18 15 AM](https://github.com/user-attachments/assets/fddfbaed-a927-4b82-b348-9ded781d3b5e)

2. tables arent deployed by the dev command:
![Screenshot 2025-01-27 at 10 19 42 AM](https://github.com/user-attachments/assets/ddbf7e88-9945-4da4-9410-6c604c116aa3)

pascal-botpress commented on packages/sdk/src/plugin/definition.ts:
```diff
@@ -62,6 +66,9 @@ export type PluginDefinitionProps<
     [K in keyof TActions]: ActionDefinition<TActions[K]>
   }
+  tables?: {
+    [K in Extract<keyof TTables, string>]: TableDefinition<TTables[K]>
```
I was having some issues with typescript where it tried to add symbol indexes (not just string indexes) and it messed with other types, so I pretty much added Extract everywhere to solve it, but once it's ready for review, I'll try removing the Extracts and see what breaks

pascal-botpress committed:
allow filtering on generated columns

pascal-botpress committed:
redefine BaseTable

pascal-botpress committed:
move parsing-utils to src/tables

pascal-botpress committed:
don't use sdk.JSONSchemaOfType

pascal-botpress committed:
override the default zui parse error message

pascal-botpress committed:
add tests for AtLeastOne, AtLeastOneProperty, ExactlyOneProperty

pascal-botpress committed:
codegen: add tables to plugins

pascal-botpress committed:
remove extraneous Extract's

pascal-botpress commented on packages/cli/src/command-implementations/deploy-command.ts:
```diff
@@ -348,10 +348,117 @@ export class DeployCommand extends ProjectCommand<DeployCommandDefinition> {
+
+      this.logger.log(`Deploying table "${tableName}"...`)
+
+      if (existingTable) {
```
In my first revision, I used to have a check in place for removed tables, but I removed it because the user might not want to define _all_ of their tables as-code and if you have 5 tables and your bot-as-code only uses one, it could spam the console. I could perhaps simply display a message saying "Table X is missing from your bot definition and will be ignored", without prompting for Y/N and doing deletions

pascal-botpress committed:
add a message about undefined tables

pascal-botpress committed:
move table publishing to its own file, reuse for bp dev

pascal-botpress committed:
fix typing regression in TableRowColumnFilters

Somehow, removing the Extract<> made it so that we can disregard
column typings and pass any value types when doing operations
on tables.

franklevasseur reviewed (approved):
I tried it, it's very good. Let's iterate on that

pascal-botpress committed:
add linter rules for tables
